Implement GET apis/v2beta1/experiments/{exp_id}/runs

difince opened the pull request:
- Implement GET apis/v2beta1/experiments/{exp_id}/runs
- add validation if the experiment exists
- api: rename ListExperiment to ListExperiments
- fix:generator.sh: add missing run.swagger.json
- move ValidateListOptions under `list` package
- move isAuthozired func under resourceManager.go
- exported ValidatedListOptions and IsAuthorized funcs are now used by both v1 and v2

Signed-off-by: Diana Atanasova <dianaa@vmware.com>

**Description of your changes:**


**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 
<!--
   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.
-->

difince committed:
Implement GET apis/v2beta1/experiments/{exp_id}/runs

- Implement GET apis/v2beta1/experiments/{exp_id}/runs
- Export ValidatedListOptions and IsAuthorized funcs (used from v1 and v2)
- api: rename ListExperiment to ListExperiments
- fix:generator.sh: add missing run.swagger.json

Signed-off-by: Diana Atanasova <dianaa@vmware.com>

difince committed:
Remove unused functions related to resource_references

Signed-off-by: Diana Atanasova <dianaa@vmware.com>

annajung commented on backend/src/apiserver/server/list_request_util.go:
```diff
@@ -153,7 +153,7 @@ func parseAPIFilter(encoded string) (*api.Filter, error) {
 }
 
-func validatedListOptions(listable list.Listable, pageToken string, pageSize int, sortBy string, filterSpec string) (*list.Options, error) {
+func ValidatedListOptions(listable list.Listable, pageToken string, pageSize int, sortBy string, filterSpec string) (*list.Options, error) {
```
I think we should think a bit more about how we want to structure and reuse components from v1.

- Because v2beta is a separate directory under `/src/apiserver`, I am concerned that it gives a false impression that files outside of the directory are not used as all other files are related to v1
- If we start using the v1 component for v2 without some structure in place, I think it would complicate v1 clean-up in the future

Would it make more sense to duplicate the file under `/src/apiserver/v2beta1/server` or introduce a common directory between the two APIs?

annajung replied:
With differences in packages `v2beta1` and `server`, we should look into the best way to handle the deprecation of `server` in the future

difince replied:
Yes, we need to think about how to handle the common code appropriately. 
I do not have a strong opinion but I tend to be not in favor of duplicating code.. as if there is a bug or need for enhancement, it should be done in both places. And if a bug is opened for just one of the versions - the developer may not realize that the fix should be done in two places.  In addition, unit tests should be also duplicated. V1 could be held for a long time - 1 year until removed. 

- So, I am more in favor of making v2 dependent on some common exported v1.funcs, but the opposite should be forbidden. Once it comes the time to remove v1, IMO if I delete  v1.func that is used in v2 - I will have compilation errors -> so I could easily keep just what I need. Another approach is to inspect the imports....  

- Or to make a restructure -  create a dedicated folder for v1. Downside - Git Blame with annotate won't be informative ... as the files will be considered new
```
|> server
   |> v1
   |> v2beta1 
   | utils.go
   | utils_test.go
   | common.go
   | common_test.go
```  
WDYT?

difince replied:
Another idea - for commonly used funcs we may use the [common](https://github.com/difince/pipelines/pull/3/commits/9482f5053275d560df9efd8063c0f01a638ed4a0) package that already exists.

For `validatedListOptions` it looks there is a better place - the [list](https://github.com/difince/pipelines/tree/master/backend/src/apiserver/list) folder

difince replied:
[Commit ](https://github.com/difince/pipelines/pull/3/commits/f5b501c908729a2ebad0c659e0d9a8acd1a8bd35) moves `validatedListOptions` into the `list` package.

annajung replied:
I like the idea of moving the functions into suitable packages and we should consider https://github.com/difince/pipelines/pull/3#discussion_r920119937 before deciding on where it might be the best fit

Linchin replied:
> I like the idea of moving the functions into suitable packages and we should consider [#3 (comment)](https://github.com/difince/pipelines/pull/3#discussion_r920119937) before deciding on where it might be the best fit

Did you mean you prefer the second method where we have separate folders for v1 and v2? I like it too since the structure is a lot clearer. From [this discussion](https://github.com/difince/pipelines/pull/3#discussion_r920402292) I think we will need some new functions for resource_manager.go, which can be put inside the existing files, or a new file. For *_storage.go we might just need to make small changes to the func.

difince committed:
Move IsAuthorize func into resourceManager

Move IsAuthorize func into resourceManager and make
resourceManager.AuthenticateRequest private.

resourcemanager.IsRequestAuthorized shoud be made private
as well, but first https://github.com/kubeflow/pipelines/pull/7351/files
issue need to be fixed.

Signed-off-by: Diana Atanasova <dianaa@vmware.com>

difince commented on backend/src/apiserver/resource/resource_manager.go:
```diff
@@ -1274,7 +1274,45 @@ func (r *ResourceManager) GetPipelineVersionTemplate(versionId string) ([]byte,
+// can perform some action (verb) on a resource (resourceType/resourceName) living in the
+// target namespace. If the returned error is nil, the authorization passes. Otherwise,
+// authorization fails with a non-nil error.
+func (r *ResourceManager) IsAuthorized(ctx context.Context, resourceAttributes *authorizationv1.ResourceAttributes) error {
```
@annajung, take a look into my last [commit](https://github.com/difince/pipelines/pull/3/commits/9482f5053275d560df9efd8063c0f01a638ed4a0). 
After I think twice I got that `IsAuthorized` func better fit in `ResourceManager`. The idea is that  `IsAuthorized` to be public func while [authenticateRequests](https://github.com/difince/pipelines/blob/9482f5053275d560df9efd8063c0f01a638ed4a0/backend/src/apiserver/resource/resource_manager.go#L1315) and [IsRequestAuthorized](https://github.com/difince/pipelines/blob/9482f5053275d560df9efd8063c0f01a638ed4a0/backend/src/apiserver/resource/resource_manager.go#L1333) to be kept private. 

Note:  This [issue](https://github.com/kubeflow/pipelines/pull/7351/files) prohibits transforming `IsRequestAuthorized` into a private func as IsRequestAuthorized is directly used when Uploading Pipeline, but it shouldn't

difince replied:
Another even better place for these three funcs we could consider in the future is the `/auth` package...

difince committed:
move validateListOptions fun into `list` package

Move the func to the package it really belongs to.

Signed-off-by: Diana Atanasova <dianaa@vmware.com>

annajung commented on backend/src/apiserver/list/list_request_util.go:
```diff
@@ -0,0 +1,71 @@
+// parseAPIFilter attempts to decode a url-encoded JSON-stringified api
+// filter object. An empty string is considered valid input, and equivalent to
+// the nil filter, which trivially does nothing.
+func parseAPIFilter(encoded string) (*api.Filter, error) {
```
In this approach, are you suggesting that this doesn't need to use the `apiv2beta1`?

difince replied:
No, I don't. 
Now,  `parseAPIFilter`  being part of `list` package is used by both v1 and v2 via  [ValidatedListOptions](https://github.com/difince/pipelines/blob/a111784eff679ffc76d57f7d2920e1ee0794dfec/backend/src/apiserver/list/list_request_util.go#L36-L43)

annajung replied:
As discussed offline, we need `parseAPIFilter` to return `apiv2beta1.Filter` so that it's not dependent on the code generated from v1 proto files. 

Therefore, the concerns we are trying to address are
- How do we want to handle common proto files that do not change between v1 and v2. Do we want to duplicate them or look into separating them out into common / v1 / v2?
- How do we handle functions like `parseAPIFilter` that depend on the (common / v1) proto files. In this case, we need to have the same exact function with different return types (`*api.Filter` vs `*apiv2beta1.Filter`). One way to go about it is to add an if/else statement or another way is to duplicate the code for v2 (duplicate in the same file or a new file). 
  - Note: while some are simple as if/else, v2 changes the parameter type and removes resource references. Therefore, some functions may require complete re-write

difince replied:
To add on Anna, we need to carefully handle:
- All POST ResourceManager methods, as they accept as a param v1 specific api Object s
- Need to appropriately handle all defined in *.proto files EMUNs, as they are used everywhere including *_storage.go files.

Linchin replied:
So not only does server package need to be changed according to v2 api, but also resource manager and storage. We need to consider v1 and v2 separation on all three levels.

Linchin replied:
> How do we want to handle common proto files that do not change between v1 and v2. Do we want to duplicate them or look into separating them out into common / v1 / v2?

It feels like proto files can be separated between v1 and v2, even if some files may be duplicated. They are essentially two sets of APIs. 

> * How do we handle functions like parseAPIFilter that depend on the (common / v1) proto files. In this case, we need to have the same exact function with different return types (*api.Filter vs *apiv2beta1.Filter). One way to go about it is to add an if/else statement or another way is to duplicate the code for v2 (duplicate in the same file or a new file).
>     * Note: while some are simple as if/else, v2 changes the parameter type and removes resource references. Therefore, some functions may require complete re-write

If the function requires a re-write, we will have to write a separate function. As to the ones solvable by a simple if/else, both ways are ok with me, since future removal of v1 won't be of much work either way. 


> To add on Anna, we need to carefully handle:
> * All POST ResourceManager methods, as they accept as a param v1 specific api Object s

Looks like the following POST functions in `resource_manager.go` took in v1 specific api objects, not counting the unexposed functions:
* CreateRun()
* CreateTask()
* CreateJob()
* ReportMetric()
* CreatePipelineVersion()

A quick look at these functions gives me the impression that these v1 api objects are quite deeply ingrained in these functions. We might have to write separate functions for v2, as well. 

> * Need to appropriately handle all defined in *.proto files EMUNs, as they are used everywhere including *_storage.go files.

Did you mean ENUMs, or am I missing something? Could you please show me an example where *.proto are used in *_storage.go files? As far as I can see, all input params are all model objects, which are not affected by *.proto. I do see several constants defined in api being used here ([example](https://github.com/difince/pipelines/blob/127dab4d4671849d596c455f4619ea807d22f6ea/backend/src/apiserver/storage/run_store.go#L355)), but as they are not lots of code, maybe some if/else branch will be sufficient?


These are just my preliminary thoughts. WDYT?

difince replied:
> Did you mean ENUMs, or am I missing something? Could you please show me an example where *.proto are used in *_storage.go files? As far as I can see, all input params are all model objects, which are not affected by *.proto. I do see several constants defined in api being used here ([example](https://github.com/difince/pipelines/blob/127dab4d4671849d596c455f4619ea807d22f6ea/backend/src/apiserver/storage/run_store.go#L355)), but as they are not lots of code, maybe some if/else branch will be sufficient?

yes, I meant ENUMS. The example you provide comes from an ENUM which is defined [here](https://github.com/difince/pipelines/blob/5b5f320736064a62daf4f527f58512c8ec5b1d32/backend/api/v2beta1/run.proto#L235-L238) . In v2 we do not change the ENUMs. What could be the condition of the if clause .. ?  And it would be strange having "if"  and "else" that are actually equivalent...

Linchin replied:
> yes, I meant ENUMS. The example you provide comes from an ENUM which is defined [here](https://github.com/difince/pipelines/blob/5b5f320736064a62daf4f527f58512c8ec5b1d32/backend/api/v2beta1/run.proto#L235-L238) . In v2 we do not change the ENUMs. What could be the condition of the if clause .. ? And it would be strange having "if" and "else" that are actually equivalent...

Ah I see! In the example that I linked above, the line `else if r.StorageState != api.Run_STORAGESTATE_AVAILABLE.String()` has to be altered so `r.StorageState` can be compared against both `apiv1beta1.Run_STORAGESTATE_AVAILABLE.String()` and `apiv1beta2.Run_STORAGESTATE_AVAILABLE.String()` and decide whether v1 or v2 it came from. That's what I meant.

By the way, as a non-goal for now, maybe we can consider decoupling `resource_manage` from APIs, and further, decouple the `storage` from APIs. This means we convert the API objects into model objects on the server level, rather than converting at resource manager level, as the code is now.

difince committed:
LusRuns by experimentId: add validation if experiment exists

Signed-off-by: Diana Atanasova <dianaa@vmware.com>

difince reviewed (commented):
Closing it as not relevant anymore
